kv/txn_coord_sender.go

```diff
@@ -409,266 +389,85 @@ func (tc *TxnCoordSender) sendOne(ctx context.Context, call proto.Call) {
-			// This is expected: the Range doesn't bother copying unless the
-			// object changes.
-			respHeader.Txn = gogoproto.Clone(header.Txn).(*proto.Transaction)
-		}
-		tc.updateResponseTxn(header, respHeader)
-	}
-
-	if txn := call.Reply.Header().Txn; txn != nil {
-		if !header.Txn.Equal(txn) {
-			panic("transaction ID changed")
-		}
-		tc.Lock()
-		txnMeta := tc.txns[id]
-		// If this transactional command leaves transactional intents, add the key
-		// or key range to the intents map. If the transaction metadata doesn't yet
-		// exist, create it.
-		if call.Reply.Header().GoError() == nil {
-			if proto.IsTransactionWrite(call.Args) {
-				if txnMeta == nil {
-					txn.Writing = true
-					trace.Event("coordinator spawns")
-					txnMeta = &txnMetadata{
-						txn:              *txn,
-						keys:             cache.NewIntervalCache(cache.Config{Policy: cache.CacheNone}),
-						firstUpdateNanos: tc.clock.PhysicalNow(),
-						lastUpdateNanos:  tc.clock.PhysicalNow(),
-						timeoutDuration:  tc.clientTimeout,
-						txnEnd:           make(chan struct{}),
-					}
-					tc.txns[id] = txnMeta
-					if !tc.stopper.RunAsyncTask(func() {
-						tc.heartbeatLoop(id)
-					}) {
-						// The system is already draining and we can't start the
-						// heartbeat. We refuse new transactions for now because
-						// they're likely not going to have all intents committed.
-						// In principle, we can relax this as needed though.
-						call.Reply.Header().SetGoError(&proto.NodeUnavailableError{})
-						tc.Unlock()
-						tc.unregisterTxn(id)
-						return
-					}
-				}
-				txnMeta.addKeyRange(header.Key, header.EndKey)
-			}
-			// Update our record of this transaction.
-			if txnMeta != nil {
-				txnMeta.txn = *txn
-				txnMeta.setLastUpdate(tc.clock.PhysicalNow())
-			}
-		}
-		tc.Unlock()
-	}
+	// Send the command through wrapped sender, taking appropriate measures
+	// on error.
+	var br *proto.BatchResponse
+	{
+		var err error
+		br, err = tc.wrapped.SendBatch(ctx, ba)
 
-	// Cleanup intents and transaction map if end of transaction.
-	switch t := call.Reply.Header().GoError().(type) {
-	case *proto.TransactionStatusError:
-		// Likely already committed or more obscure errors such as epoch or
-		// timestamp regressions; consider it dead.
-		tc.cleanupTxn(trace, t.Txn)
-	case *proto.TransactionAbortedError:
-		// If already aborted, cleanup the txn on this TxnCoordSender.
-		tc.cleanupTxn(trace, t.Txn)
-	case *proto.OpRequiresTxnError:
-		// Run a one-off transaction with that single command.
-		if log.V(1) {
-			log.Infof("%s: auto-wrapping in txn and re-executing", call.Method())
-		}
-		// TODO(tschottdorf): this part is awkward. Consider resending here
-		// without starting a new call, which is hard to trace. Plus, the
-		// below depends on default configuration.
-		tmpDB, err := client.Open(
-			fmt.Sprintf("//?priority=%d",
-				call.Args.Header().GetUserPriority()),
-			client.SenderOpt(tc))
+		err = tc.updateState(ctx, ba, br, err)
```

Comment:
That's true, will do that.
On Aug 27, 2015 21:06, "Tamir Duberstein" <notifications@github.com> wrote:

> In kv/txn_coord_sender.go
> <https://github.com/cockroachdb/cockroach/pull/2141#discussion_r38134062>:
>
> > -	case *proto.TransactionAbortedError:
> > -		// If already aborted, cleanup the txn on this TxnCoordSender.
> > -		tc.cleanupTxn(trace, t.Txn)
> > -	case *proto.OpRequiresTxnError:
> > -		// Run a one-off transaction with that single command.
> > -		if log.V(1) {
> > -			log.Infof("%s: auto-wrapping in txn and re-executing", call.Method())
> > -		}
> > -		// TODO(tschottdorf): this part is awkward. Consider resending here
> > -		// without starting a new call, which is hard to trace. Plus, the
> > -		// below depends on default configuration.
> > -		tmpDB, err := client.Open(
> > -			fmt.Sprintf("//?priority=%d",
> > -				call.Args.Header().GetUserPriority()),
> > -			client.SenderOpt(tc))
> > +		err = tc.updateState(ctx, ba, br, err)
>
> right, but you can if err := ...; err != nil {
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2141/files#r38134062>.
>